src/sinks/datadog_archives.rs

```diff
@@ -0,0 +1,571 @@
+    lazy_static! {
+        static ref KEY_PREFIX: Template =
+            Template::try_from("/dt=%Y%m%d/hour=%H/").expect("invalid object key format");
+    }
+    let key = KEY_PREFIX
+        .render_string(&event)
+        .map_err(|error| {
+            emit!(TemplateRenderingFailed {
+                error,
+                field: Some("key_prefix"),
+                drop_event: true,
+            });
+        })
+        .ok()?;
+
+    let mut log = event.into_log();
+
+    log.insert("_id", generate_log_id());
+
+    let timestamp = log
+        .remove(crate::config::log_schema().timestamp_key())
+        .unwrap_or_else(|| chrono::Utc::now().timestamp_millis().into());
+    log.insert(
+        "date",
+        timestamp
+            .as_timestamp()
+            .cloned()
+            .unwrap_or_else(chrono::Utc::now)
+            .to_rfc3339_opts(SecondsFormat::Millis, true),
+    );
+    if let Some(message) = log.remove(crate::config::log_schema().message_key()) {
+        log.insert("message", message);
+    }
+    if let Some(message) = log.remove(crate::config::log_schema().host_key()) {
+        log.insert("host", message);
+    }
+
+    let mut attributes = BTreeMap::new();
+    let custom_attributes: Vec<String> = log
+        .keys()
+        .filter(|path| !RESERVED_ATTRIBUTES.contains(path.as_str()))
+        .collect();
+    for path in custom_attributes {
+        if let Some(value) = log.remove(&path) {
+            attributes.insert(path, value);
+        }
+    }
+    log.insert("attributes", attributes);
+
+    let bytes = serde_json::to_vec(&log)
+        .map(|mut b| {
+            b.push(b'\n');
+            b
+        })
+        .expect("Failed to encode event as json, this is a bug!");
+
+    Some(EncodedEvent {
+        item: PartitionInnerBuffer::new(bytes, key.into()),
+        finalizers: log.metadata_mut().take_finalizers(),
+    })
+}
+
+/// Generates a unique ID compatible with DD
+fn generate_log_id() -> String {
+    let mut id = BytesMut::with_capacity(18);
+    // timestamp in millis - 6 bytes
+    let now = Utc::now();
+    id.put_int(now.timestamp_millis(), 6);
+    // namespace + version, both 0 - 2 bytes
+    id.put_i16(0);
+
+    // host-based number - 5 bytes
+    lazy_static! {
+        static ref HOST_NUMBER: Vec<u8> = host_number();
+    }
+    id.put_slice(&HOST_NUMBER[..]);
+
+    // counter - 5 bytes
+    id.put_u8(0);
+    // one padding byte
+    // 4 bytes for the counter should be more than enough - it should be unique for 1 millisecond only
+    lazy_static! {
+        static ref COUNTER: CounterU32 = CounterU32::new(0);
+    }
+    id.put_u32(COUNTER.inc()); // 4 bytes
+
+    base64::encode(id.freeze())
+}
+
+/// Returns 5 first bytes of the MAC address XOR-ed with random bytes
+fn host_number() -> Vec<u8> {
+    extern crate mac_address;
+    use mac_address::get_mac_address;
+    use rand::Rng;
+
+    let mac = get_mac_address().unwrap().unwrap().bytes();
```

Comment:
The MAC is being used in this PR to make a unique host identifier, or unique-ish?

Reply:
The idea here was to follow the [Snowflake ID](https://en.wikipedia.org/wiki/Snowflake_ID) and follow the similar approach used by DD, though there is no hard requirements to use MAC-address - it is being used to avoid collisions in generated IDs, but I think any unique-ish number would be sufficient. Do you have any concerns around using it?

Reply:
Got it. No, I ask because I wasn't clear on the goal here. Rather than in-lining the concept of snowflake id in this PR consider making another, smaller PR  to add this to vector-core. Seems like there'd be general utility.

Reply:
On a related note, are we sure we want to `unwrap` here? At the very least, it would be nice to see `expect` with a reason if we are OK panicing.

Reply:
Oh yeah, no we aren't. If it's not possible to make this kind of ID without access to the MAC address we should prefer something non-fallible.

Reply:
thanks for your comments! I've made this function non-fallible(fallback to random bytes if MAC isn't available).
With regards to making ID generation a general function - this is not exact Snowflake ID implementation, just follows a similar approach and it is very DD-specific(i.e. we do have some constraints on this field). Also I don't see atm any other places in the future where we could use it even for DD-specific functionality, so I'd keep it here for now if that's ok with you.

Reply:
>  I've made this function non-fallible(fallback to random bytes if MAC isn't available).

Did you forget to push a commit? There's still a couple of unwraps here. 

> With regards to making ID generation a general function - this is not exact Snowflake ID implementation, just follows a similar approach and it is very DD-specific(i.e. we do have some constraints on this field). Also I don't see atm any other places in the future where we could use it even for DD-specific functionality, so I'd keep it here for now if that's ok with you.

Sure, that's fine. This function needs to be better documented -- it's not clear what the form of the ID is supposed to be, what its constraints are other than what's implicit in the code -- and if we only use the MAC to get 5 bytes which we then randomly XOR I question why we need the MAC  in the first place. It's hard to know for sure since I don't know if this host number is meant to be stable across restarts -- seemingly not -- or what the consequence of collision is for the API.

Reply:
right, sorry, forgot to push this commit. More documentation is in the [RFC](https://github.com/DataDog/vector-enterprise/blob/54965df4bc2958f92caa1fdbbccdb8997adcf2a3/rfcs/2021-07-23-172-rehydration-from-vector-archives.md#generation-of-unique-log-ids-required-to-index-rehydrated-logs) I'm just wondering how much we should document and expose DD-specific internals. @binarylogic maybe you have any thoughts? 
In general collision means these events will be seen as duplicates, so only the last of them would be indexed, also they do not need to be stable across restarts - it's only to prevent collisions.
Why we need XOR - from what I can tell seems like we do it for security reasons to avoid exposing an original MAC address, though not sure what are these concerns exactly. 
Overall, using MAC address is not a hard requirement, the only ones are:
- 18 bytes
- first 6 bytes is a timestamp in millis
- then some unique(for a given timestamp) number
Considering this I'm ok with using just a random sequence of bytes instead of the munged MAC address. 
WDYT?

Reply:
There needs to be enough documentation here to tell incidental behavior from intentional. There's good material in these PR comments; it just needs to be landed in-code. 

With regard to the MAC address, it seems like you can drop that dependency entirely. There's enough entropy in a `u32` for disambiguation purposes.

Reply:
Sure, will update the comments 👍  
>With regard to the MAC address, it seems like you can drop that dependency entirely. There's enough entropy in a u32 for disambiguation purposes.

I'm afraid we still need some machine-dependent part of the ID(or at least unique-ish across multiple hosts). Looking at some other implementations I think using `md5(hostname)` like [here](https://github.com/rs/xid/blob/92f87d0607996d5a16f94a308e0bff833d3ba1dc/id.go#L117)

Reply:
It's not clear to me why we need anything machine-dependent based on the requirements you've laid out, other implementations notwithstanding. As I understand it a host number is:

* a disambiguation tool,
* a 5-byte vector of `u8`
* needs enough entropy to avoid but not eliminate collisions (else the byte space would be bigger) and
* does not persist for a host across vector restarts.

Am I missing something? If those are all true there's no need for anything but a random 5 bytes from a source of true randomness, which `thread_rng` is.

Reply:
Yes, you are right, and agree with you. Ok, I will just rely on the entropy of the random bytes and document all constraints.

Reply:
Updated this part.